### PR TITLE
plastid metagene_generate, make_wiggle, psite

### PR DESCRIPTION
**Pull request by suhrig**
This PR adds plastid with three submodules:
- metagene_generate
- psite
- make_wiggle

These functions are required to compute p-sites from Ribo-seq data.

I have also added test data here: https://github.com/nf-core/test-datasets/pull/1809

## PR checklist

- [X] This comment contains a description of changes (with reason).
- [X] If you've fixed a bug or added code that should be tested, add tests!
- [X] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [X] If necessary, include test data in your PR.
- [X] Remove all TODO statements.
- [X] Emit the `versions.yml` file.
- [X] Follow the naming conventions.
- [X] Follow the parameters requirements.
- [X] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [X] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [X] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`

### CODE REVIEW

File: modules/nf-core/plastid/psite/main.nf [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    val min_length
+    val max_length
+    val require_upstream
+    val default_offset
+
+    output:
+    tuple val(meta), path("*_metagene_profiles.txt"), emit: metagene_profiles
+    tuple val(meta), path("*_p_offsets.png")        , emit: p_offsets_png
+    tuple val(meta), path("*_p_offsets.txt")        , emit: p_offsets
+    path "versions.yml"                             , emit: versions
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/psite/main.nf:19]: Might be sensible to use the new topics syntax instead

Comment (pinin4fjords) [reply modules/nf-core/plastid/psite/main.nf:19]: https://github.com/nf-core/modules/blob/c93d7d10f9e67e868d27073e600114ae0ee1adef/modules/nf-core/ribodetector/main.nf#L17

File: modules/nf-core/plastid/psite/main.nf [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    label "process_single"
+
+    conda "${moduleDir}/environment.yml"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/plastid:0.6.1--py39had3e4b6_2':
+        'biocontainers/plastid:0.6.1--py39had3e4b6_2' }"
+
+    input:
+    tuple val(meta), path(bam), path(bam_index)
+    path rois_txt
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/psite/main.nf]: Pretty much everything should have metas these days

File: modules/nf-core/plastid/psite/main.nf [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    script:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def args = task.ext.args ?: ""
+    if (require_upstream) args += "--require_upstream"
+    """
+    psite \
+        "$rois_txt" \\
+        "$prefix" \\
+        --min_length "$min_length" --max_length "$max_length" \\
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/psite/main.nf]: The are optional, no? those should normally come via ext.args.

File: modules/nf-core/plastid/psite/main.nf [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    tuple val(meta), path("*_p_offsets.txt")        , emit: p_offsets
+    path "versions.yml"                             , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def args = task.ext.args ?: ""
+    if (require_upstream) args += "--require_upstream"
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/psite/main.nf]: Seems like another optional thing that should come via ext.args

File: modules/nf-core/plastid/psite/main.nf [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    touch ${prefix}_metagene_profiles.txt
+    touch ${prefix}_p_offsets.png
+    touch ${prefix}_p_offsets.txt
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        plastid: 0.6.1
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/psite/main.nf]: Does the tool not have a version command? We don't like to hard code these

File: modules/nf-core/plastid/psite/main.nf [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    sed -i '/^##/d' *.txt # remove variable comment header
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        plastid: 0.6.1
+    END_VERSIONS
+    """
+
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/psite/main.nf:42]: The stub should reflect all of the handling logic from the main script

File: modules/nf-core/plastid/metagene_generate/main.nf [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    tag "$gtf"
+    label "process_low"
+
+    conda "${moduleDir}/environment.yml"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/plastid:0.6.1--py39had3e4b6_2':
+        'biocontainers/plastid:0.6.1--py39had3e4b6_2' }"
+
+    input:
+    path gtf
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/metagene_generate/main.nf]: You definitely need a meta, and you probably want to bundle these two inputs together (unless the val is at all optional, in which it should come via ext.args)

File: modules/nf-core/plastid/metagene_generate/main.nf [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    val landmark
+
+    output:
+    path "*_rois.txt"  , emit: rois_txt
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/metagene_generate/main.nf]: Outputs will need the metas too

File: modules/nf-core/plastid/metagene_generate/main.nf [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    path "*_rois.bed"  , emit: rois_bed
+    path "versions.yml", emit: versions
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/metagene_generate/main.nf:17]: Again, can probably do with topics

File: modules/nf-core/plastid/metagene_generate/main.nf [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    """
+    metagene generate "${gtf.baseName}" --landmark "$landmark" --annotation_files "$gtf" $args
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/metagene_generate/main.nf]: Are any of these optional? Would `metagene generate` run without them? e.g. 'landmark'? If so, that should go via ext.args

File: modules/nf-core/plastid/metagene_generate/main.nf [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    sed -i '/^##/d' *_rois.* # remove variable comment header
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        plastid: 0.6.1
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/metagene_generate/main.nf]: again with the hard code

Comment (pinin4fjords) [reply modules/nf-core/plastid/metagene_generate/main.nf]: if it needs to be a static version it should go like: https://nf-co.re/docs/guidelines/components/modules#emission-of-versions

File: modules/nf-core/plastid/make_wiggle/main.nf [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    label "process_single"
+
+    conda "${moduleDir}/environment.yml"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/plastid:0.6.1--py39had3e4b6_2':
+        'biocontainers/plastid:0.6.1--py39had3e4b6_2' }"
+
+    input:
+    tuple val(meta), path(bam), path(bam_index), path(p_offsets)
+    val min_length
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/make_wiggle/main.nf]: again, just checking, but if a `val` argument isn't absolutely required for the command operation it should come via ext.args.

File: modules/nf-core/plastid/make_wiggle/main.nf [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    path "versions.yml", emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def extension = output_format == "bedgraph" ? "bedgraph" : "wig"
+    def args = task.ext.args ?: ""
+    if (fiveprime_variable) args += "--fiveprime_variable"
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/make_wiggle/main.nf]: I suspect we won't need this with good ext.args handling

File: modules/nf-core/plastid/make_wiggle/main.nf [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    make_wiggle \\
+        --count_files "$bam" \\
+        --offset "$p_offsets" \\
+        --min_length "$min_length" \\
+        --output_format "$output_format" \\
+        --verbose \\
+        -o "$prefix" \\
+        $args
+
+    sed -i '/^##/d' *.wig # remove variable comment header
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/make_wiggle/main.nf]: Let's not edit command output. If you're doing this for the snapshot purposes there are other workarounds (like checking for content rather than a full snapshot)

File: modules/nf-core/plastid/make_wiggle/main.nf [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+    if [ "$output_format" = "bedgraph" ]; then
+        for FILE in *.wig; do
+            mv "\$FILE" "\${FILE%.wig}.bedgraph"
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/make_wiggle/main.nf:41]: Why the rename?

File: modules/nf-core/plastid/make_wiggle/main.nf [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def extension = output_format == "bedgraph" ? "bedgraph" : "wig"
+    """
+    touch ${prefix}_fw.$extension
+    touch ${prefix}_rc.$extension
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        plastid: 0.6.1
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/make_wiggle/main.nf]: ditton on the version comment for other modules

File: modules/nf-core/plastid/make_wiggle/main.nf (hunk 15 above)
Comment (suhrig) [reply modules/nf-core/plastid/make_wiggle/main.nf:41]: The tool names the files .wig, even when bedgraph is chosen as output format. I don't mind leaving it like this, but it feels wrong and I thought it could lead to confusion with downstream tools that determine the file type from the extension. What do you recommend?

**Comment by suhrig** [comment]
Thanks for the comprehensive feedback! I will address the points and change the code accordingly.

File: modules/nf-core/plastid/psite/main.nf (hunk 5 above)
Comment (suhrig) [reply modules/nf-core/plastid/psite/main.nf]: I searched. It doesn't. I will use the `VERSION` variable workaround with the warning comment.

File: modules/nf-core/plastid/make_wiggle/main.nf (hunk 15 above)
Comment (pinin4fjords) [reply modules/nf-core/plastid/make_wiggle/main.nf:41]: I see, OK, guess that's fair.

File: modules/nf-core/plastid/psite/main.nf (hunk 2 above)
Comment (suhrig) [reply modules/nf-core/plastid/psite/main.nf]: I have added the meta to all input/output files.

File: modules/nf-core/plastid/psite/main.nf (hunk 4 above)
Comment (suhrig) [reply modules/nf-core/plastid/psite/main.nf]: All optional parameters have been removed.

File: modules/nf-core/plastid/make_wiggle/main.nf (hunk 13 above)
Comment (suhrig) [reply modules/nf-core/plastid/make_wiggle/main.nf]: This is the only parameter that is mandatory. `make_wiggle` accepts p-site offsets to be supplied in two ways: as a static integer value or as a file listing a different value for each read length. Unfortunately, in both cases the argument is supplied via the parameter `--offset`. Nextflow cannot deal with the different data types. It has to be either a val() or a path(). I decided to go for path(), because I believe this is more commonly used and also more flexible. Users who wish to use a static value for all read lengths can still supply a file with just one entry `default <some integer>`. When a path() is used, then the parameter `--fiveprime_variable` is mandatory.

File: modules/nf-core/plastid/make_wiggle/main.nf (hunk 14 above)
Comment (suhrig) [reply modules/nf-core/plastid/make_wiggle/main.nf]: I'm using content-based checks now.

File: modules/nf-core/plastid/psite/main.nf (hunk 1 above)
Comment (suhrig) [reply modules/nf-core/plastid/psite/main.nf:19]: As discussed on Slack I will stick with the traditional method of collecting versions. I hope that's okay.

**Comment by suhrig** [comment]
Thanks again for your feedback. All issues should be resolved now. Let me know if you would like any further changes.

File: modules/nf-core/plastid/make_wiggle/main.nf [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    script:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def extension = output_format == "bedgraph" ? "bedgraph" : "wig"
+    def args = task.ext.args ?: ""
+    def VERSION = "0.6.1" // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
+    """
+    make_wiggle \\
+        --count_files "$bam" \\
+        --offset "$p_offsets" \\
+        --fiveprime_variable \\
```
Comment (pinin4fjords) [inline modules/nf-core/plastid/make_wiggle/main.nf]: You're still hard-coding things that look optional (correct me if I'm wrong). These should go via ext.args, and you can do that via nextflow.config files in the test directory and the appropriate `withName` selectors. There's lots of examples of this in the modules repo if you look about.

Comment (suhrig) [reply modules/nf-core/plastid/make_wiggle/main.nf]: `make_wiggle` doesn't run when `--fiveprime_variable` is absent and the argument to `--offset` is  a file. I can move `--fiveprime_variable` to `ext.args` if you like, but this will mean that anyone using the module will have to do this.

Comment (pinin4fjords) [reply modules/nf-core/plastid/make_wiggle/main.nf]: I've opened a PR to your branch with some nf-core compliance tweaks by way of illustration: https://github.com/suhrig/modules/pull/1

Since 'a' mapping rule is mandatory (one of `--fiveprime`, `--threeprime`, `--center`, `--fiveprime_variable`), but `--fiveprime_variable` is not, I've added it as a mapping_rule val input with an enum. 

The p_offsets file is then only passed when mapping_rule == 'fiveprime_variable'. Moved output_format to ext.args since it's optional, but we can still respond to it for the extension.

**Comment by pinin4fjords** [comment]
Hi @suhrig, I've opened another PR to your branch with suggestions to fix a CI reproducibility issue: https://github.com/suhrig/modules/pull/2

The psite PNG has non-reproducible md5sums due to matplotlib rendering differences across conda builds, so I've excluded it from the snapshot (just checking existence instead).

Also tidied up the test assertions while I was at it - consolidated snapshots and removed snapshots of empty stub files.

**Comment by pinin4fjords** [comment]
@suhrig sorry, maybe I added some things to the snapshotting that shouldn't be there, let me fix it up
